Fix for #5 Use index in ExampleNameProvider

ghaith opened the pull request:
Added an index method to fetch the name
removed NonStoringBatchLinkableResourceStorageFacade

ghaith committed:
Fix for #5 Use index in ExampleNameProvider

Added an index method to fetch the name
removed NonStoringBatchLinkableResourceStorageFacade

riederm commented:
were you able to reproduce the issue first?

ghaith commented:
Yes, disable the non storing facade, create a spec and make a suite that
references it, or references all, you will get a null pointer exception
I'm not sure if this is the best fix for it though, I kept the extra
nodemodel access as backup in it

On Mon, 1 May 2017 at 11:20 Mathias Rieder <notifications@github.com> wrote:

> were you able to reproduce the issue first?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/riederm/Jnario/pull/8#issuecomment-298309779>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI1oWNnAdJsGOf-FlKiXsw2gumuZoOks5r1aPbgaJpZM4NMudb>
> .
>

riederm commented on plugins/org.jnario.spec/src/org/jnario/spec/scoping/SpecResourceDescriptionStrategy.xtend:
```diff
@@ -13,18 +13,27 @@ import org.jnario.scoping.JnarioResourceDescriptionStrategy
+			if (exampleGroup.targetType !== null) {	
+				val targetType = exampleGroup.targetType
+				if (!(targetType.type instanceof JvmVoid)) {
+					userData.put(EXAMPLE_TYPE,exampleGroup.targetType.simpleName)					
```
I'm not sure if we can safely call exampleGroup.getTargetType(). This has to sometimes return an unresolvable proxy? Otherwhise the original problem of the ExampleNameProvider wouldn't exist. 

I think the problem is that you write a spec:

MySpec.spec
```
describe MyClass{
}
```

MyClass.xtend
```
class MyClass{
}
```
So the ExampleGroup in MySpec.spec holds a reference targetType = MyClass,
I think the unresolvable problem happens, if MyClass was not generated yet, while you're indexing. E.g. in a cleanBuild.

I would rather move the NodeModel-approach to here, and remove it from the ExampleNameProvider. 
If we really want to make sure, we should also store the nodemodel in the storage (although it means a lot of more stuff)

I just checked, by default the node-model is not stored in the storage, but you can tell the storage facade to store the node-model as well. But its again a performance loss :( 
org.eclipse.xtext.resource.persistence.ResourceStorageFacade.setStoreNodeModel(boolean)

ghaith committed:
Issue 5 Moved node model parsing to the index

Removed the NodeModel parsing from the name provider and into the index
creation, I kept a fallback to try and parse the nodemodel in case the
index was not found.

ghaith commented:
I uploaded a second attempt, i cannot reproduce an error where the class has not been generated yet, i kept fallbacks for the nodemodel but moved the requests to the spec description factory
The original error, where a suite will get a nullpointer when referencing the non resolved spec is also not occuring anymore. Still not sure if we should keep the fallbacks or if you had any other way to test this
